### PR TITLE
Add documentation to YAML module

### PR DESCRIPTION
**Pull request by adlerhsieh**
Also add type restrictions to `#load` and `#load_all`.

### CODE REVIEW

**Comment by jhass** [comment]
Serialization is the act of converting a data structure into a specific format. Deserialization is the act of converting a specific format into an accessible data structure.

File: src/yaml/yaml.cr [hunk 1]
```diff
@@ -14,7 +15,34 @@ module YAML
...
+#
+# YAML.load(File.read("./foo.yml"))
+# #=> {"foo" => {"bar" => "foobar"}}
+# ```
+#
+# Returns the same value if *data* is not in YAML format.
+# ```crystal 
+# YAML.load("foo")
+# #=> "foo"
+# ```
```
Comment (jhass) [inline src/yaml/yaml.cr]: This simply is a valid YAML string which deserialization just happens to look the same. Try

```yml
1
```
```yml
"foo bar"
```
```yml
- 1
- "foo"
```

and so on.

File: src/yaml/yaml.cr [hunk 2]
```diff
@@ -14,7 +15,34 @@ module YAML
...
+#
+# Returns `Nil` if *data* is a blank string.
+# ```crystal 
+# YAML.load("")
+# #=> nil
```
Comment (jhass) [inline src/yaml/yaml.cr]: Again the absence of any value is a valid YAML document and just is represented with nil, compare

```
foo:
  bar:
  baz:
    -
    - quux
```

**Comment by adlerhsieh** [comment]
Updated. I've simplified the explanations and offer a more complete example.

File: src/yaml/yaml.cr [hunk 3]
```diff
@@ -1,5 +1,6 @@
 require "./*"
 
+# The YAML module serves to deserialize data from YAML format into accessible structure.
```
Comment (jhass) [inline src/yaml/yaml.cr]: Should this point to `yaml_mapping` like `JSON` does?

File: src/yaml/yaml.cr [hunk 4]
```diff
@@ -14,7 +15,33 @@ module YAML
   alias Type = String | Hash(Type, Type) | Array(Type) | Nil
   alias EventKind = LibYAML::EventType
 
-  def self.load(data)
+  # Deserializes a document from YAML format.
```
Comment (jhass) [inline src/yaml/yaml.cr]: "Yet another markup language format" sounds kind of  redundant.

File: src/yaml/yaml.cr [hunk 5]
```diff
@@ -14,7 +15,33 @@ module YAML
...
+  #
+  # ```yaml
```
Comment (jhass) [inline src/yaml/yaml.cr]: I don't think we have a YAML highlighter? Might be worth having the docs use highlightjs at some point.

Comment (adlerhsieh) [inline src/yaml/yaml.cr]: No we don't have it yet. But we can put it there for the future.

Comment (jhass) [inline src/yaml/yaml.cr]: So it doesn't break  the doc parser? good

Comment (adlerhsieh) [inline src/yaml/yaml.cr]: Yeap it still works.

**Comment by adlerhsieh** [comment]
I've rewritten the introductory part to include both `#load` and `YAML::Mapping` ways.

File: src/yaml/yaml.cr [hunk 6]
```diff
@@ -1,5 +1,38 @@
 require "./*"
 
+# The YAML module serves to deserialize data from YAML into accessible structure.
+#
+# ### Parsing with `#load`
+#
+# Deserializes a YAML document into a collection of `YAML::Type`.
+# A `Type` is a union of all possible YAML types, so specifying a type is necessary
```
Comment (jhass) [inline src/yaml/yaml.cr]: "casting to a specific type"?

File: src/yaml/yaml.cr [hunk 7]
```diff
@@ -1,5 +1,38 @@
...
+# before the value is accessible.
+# 
+# ```crystal
+# require "yaml"
+#
+# data = YAML.load("foo: bar")
+# (data as Hash)["foo"] #=> "bar"
+# ```
+#
+# ### Parsing and Generating with `YAML::Mapping`
```
Comment (jhass) [inline src/yaml/yaml.cr]: Where's the generation part then?

File: src/yaml/yaml.cr [hunk 8]
```diff
@@ -1,5 +1,38 @@
 require "./*"
 
+# The YAML module serves to deserialize data from YAML into accessible structure.
+#
+# ### Parsing with `#load`
+#
+# Deserializes a YAML document into a collection of `YAML::Type`.
```
Comment (jhass) [inline src/yaml/yaml.cr]: Actually just one, the toplevel is a `YAML::Type` too.

File: src/yaml/yaml.cr [hunk 9]
```diff
@@ -1,5 +1,38 @@
...
+# A `Type` is a union of all possible YAML types, so specifying a type is necessary
+# before the value is accessible.
```
Comment (jhass) [inline src/yaml/yaml.cr]: Well, I can call common methods like `to_s` beforehand. "practically usable" perhaps?

File: src/yaml/yaml.cr [hunk 10]
```diff
@@ -1,5 +1,38 @@
...
+# ```crystal
+# require "yaml"
+#
+# data = YAML.load("foo: bar")
+# (data as Hash)["foo"] #=> "bar"
+# ```
+#
+# ### Parsing and Generating with `YAML::Mapping`
+#
+# Use `YAML::Mapping` to define how an object is mapped to YAML.
```
Comment (jhass) [inline src/yaml/yaml.cr]: Maybe take some inspiration from http://crystal-lang.org/api/JSON.html?

File: src/yaml/yaml.cr [hunk 11]
```diff
@@ -1,5 +1,38 @@
 require "./*"
 
+# The YAML module serves to deserialize data from YAML into accessible structure.
```
Comment (kirbyfan64) [inline src/yaml/yaml.cr]: How about "The YAML module **deserializes YAML data** into **an** accessible structure" or similar?

File: src/yaml/yaml.cr [hunk 12]
```diff
@@ -23,7 +82,21 @@ module YAML
     end
   end
 
-  def self.load_all(data)
+  # Deserializes multiple documents from YAML.
```
Comment (kirbyfan64) [inline src/yaml/yaml.cr]: "Deserializes multiple YAML documents"

File: src/yaml/yaml.cr (hunk 11 above)
Comment (jhass) [inline src/yaml/yaml.cr]: Do you say HTML data or CSS data? Granted some people say XML data, but I think it's kind of redundant.

Comment (kirbyfan64) [inline src/yaml/yaml.cr]: @jhass How it redundant? It's 5 bytes shorter. :)

Comment (jhass) [inline src/yaml/yaml.cr]: Well, thinking about it I'd keep it out completely, "This module provides deserialization of YAML to native Crystal datastructures" or something.

Comment (kirbyfan64) [inline src/yaml/yaml.cr]: @jhass Yeah, that's way better than mine.

**Comment by adlerhsieh** [comment]
Updated. I've moved documentation of `YAML::Mapping` to the module itself.

File: src/yaml/mapping.cr [hunk 13]
```diff
@@ -1,4 +1,63 @@
+# The `#yaml_mapping` macro defines how an object is mapped to YAML.
+#
+# Once `YAML` is required, `#yaml_mapping` macro is included by `Object`.
```
Comment (jhass) [inline src/yaml/mapping.cr]: The require would be `"yaml"`

File: src/yaml/mapping.cr [hunk 14]
```diff
@@ -1,4 +1,63 @@
...
+# It takes `HashLiteral` as argument, in which attributes and types are defined.
```
Comment (jhass) [inline src/yaml/mapping.cr]: hash literal, the AST node is sort of an internal.

File: src/yaml/mapping.cr [hunk 15]
```diff
@@ -1,4 +1,63 @@
...
+# Once defined, `Object#from_yaml` maps properties for the class based on the parsed 
+# YAML document.
```
Comment (jhass) [inline src/yaml/mapping.cr]: "populates properties of the class from the YAML document"?

File: src/yaml/mapping.cr [hunk 16]
```diff
@@ -1,4 +1,63 @@
...
+#
+# employee = Employee.from_yaml("title: Manager\nname: John")
+# employee.title #=> "Manager"
+# employee.name  #=> "John"
+#
+# employee.name = "Jenny"
+# employee.name #=> "Jenny"
+# ```
+#
+# Attributes not mapped with `#yaml_mapping` will not be defined as properties.
```
Comment (jhass) [inline src/yaml/mapping.cr]: Sudden tense switch, "are not defined"

File: src/yaml/mapping.cr [hunk 17]
```diff
@@ -1,4 +1,63 @@
...
+# employee = Employee.from_yaml("title: Manager\nname: John")
+# employee.title #=> "Manager"
+# employee.name  #=> "John"
+#
+# employee.name = "Jenny"
+# employee.name #=> "Jenny"
+# ```
+#
+# Attributes not mapped with `#yaml_mapping` will not be defined as properties. 
+# Also, missing attributes raise error.
```
Comment (jhass) [inline src/yaml/mapping.cr]: "an error"

File: src/yaml/mapping.cr [hunk 18]
```diff
@@ -1,4 +1,63 @@
...
+#
+# ```crystal
+# employee = Employee.from_yaml("title: Manager\nname: John\nage: 30")
+# employee.age #=> undefined method 'age'
+#
+# employee = Employee.from_yaml("title: Manager")
+# #=> missing yaml attribute: name
```
Comment (jhass) [inline src/yaml/mapping.cr]: Would be good to hint the exception type?

File: src/yaml/mapping.cr [hunk 19]
```diff
@@ -1,4 +1,63 @@
...
+# ```
+# 
+# You can also define attributes for each mapped properties.
```
Comment (jhass) [inline src/yaml/mapping.cr]: Maybe explain that the previous form is a shortcut to `foo: {type: Bar}` and call these options?

File: src/yaml/mapping.cr [hunk 20]
```diff
@@ -1,4 +1,63 @@
...
+#       key: "firstname"
+#     }
+#   })
+# end
+# ```
+#
+# Available attributes: 
+#
+# * *type* (required) defines its type.
+# * *nilable* defines if a property can be `Nil`.
```
Comment (jhass) [inline src/yaml/mapping.cr]: can be a `Nil` or can be `nil`

File: src/yaml/mapping.cr [hunk 21]
```diff
@@ -1,4 +1,63 @@
...
+# * *key* defines whick key to read from a YAML document.
```
Comment (jhass) [inline src/yaml/mapping.cr]: defaults to the name of the property

File: src/yaml/mapping.cr [hunk 22]
```diff
@@ -1,4 +1,63 @@
...
+# * *converter* takes an alternate type for parsing and generation. A `#from_yaml`(`PullParser`) method is required in that class.
```
Comment (jhass) [inline src/yaml/mapping.cr]: I don't think the link is worth the odd formatting. Also mentioning that it should return an instance of what's given to `type` might be worth it.

File: src/yaml/yaml.cr [hunk 23]
```diff
@@ -1,5 +1,25 @@
...
+# ```crystal
+# require "yaml"
+#
+# data = YAML.load("foo: bar")  
+# (data as Hash)["foo"] #=> "bar"
+# ```
+#
+# ### Parsing with `Mapping`
+#
+# `Mapping` defines how an object is mapped to YAML. Mapped data are accessible
```
Comment (jhass) [inline src/yaml/yaml.cr]: is accessible

File: src/yaml/yaml.cr [hunk 24]
```diff
@@ -1,5 +1,25 @@
...
+# through methods call like *Foo#bar*. It is more type-safe and efficient.
```
Comment (jhass) [inline src/yaml/yaml.cr]: "through generated properties like" ?

**Comment by adlerhsieh** [comment]
Updated.

**Comment by adlerhsieh** [comment]
@jhass  Fine-tuned a bit just now and ready for review :)

**Comment by jhass** [comment]
Thanks!

**Pull request merged**